---
title: "HIV Oral 16S Amplicon Analysis"
author: "Scott A. Handley"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

# Project Description:
Analysis of 16S rRNA amplicons in oral samples (saliva) collected from HIV infected patients pre-ART treatment and 24 weeks after treatment initiation.

This document describes the analysis of dada2 resolved sequences analysis of dada2 resolved seuqnece variants of 70 samples; 35 samples were collected at the inititation of the study from male HIV positive patients (baseline) and 35 follow-up samples 24 weeks after the inititation of ART.

**Collaborator:**
Rachel Presti (prestir@wustl.edu)

**Other relevant documents:**
Preprocessing steps can be viewed in **HIV_oral_preprocessing.Rmd**.

**References:**
* http://f1000research.com/articles/5-1492/v1
* http://benjjneb.github.io/dada2/tutorial.htm 
* Read scaling and ADONIS scripts taken from [michberr]https://github.com/michberr: https://github.com/michberr/MicrobeMiseq/blob/master/R/
* CCA code adopted from [michberr]https://github.com/michberr: http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html


```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      fig.path="./figures/",
                      dev='pdf',
                      warning=FALSE,
                      message=FALSE)
```

```{r initiate-environment}
library("ggplot2")
packageVersion("ggplot2")
library("phyloseq")
packageVersion("phyloseq")
library("magrittr")
packageVersion("magrittr")
library("RColorBrewer")
packageVersion("RColorBrewer")
library("vegan")
packageVersion("vegan")
library("gridExtra")
packageVersion("gridExtra")
library("knitr")
packageVersion("knitr")
library("dplyr")
packageVersion("dplyr")
library("DESeq2")
packageVersion("DeSeq2")
library("plotly")
packageVersion("plotly")
library("picante")
packageVersion("picante")
library("indicspecies")
packageVersion("indicspecies")
library("corrplot")
packageVersion("corrplot")
library("microbiome")
packageVersion("microbiome")
library("ggpubr")
packageVersion("ggpubr")
library("survival")
packageVersion("survival")
library("randomForest")
packageVersion("randomForest")
library("HH")
packageVersion("HH")

```
Load phyloseq data generated from dada2

```{r initiate-data}
# Load Phyloseq Object generated in HIV_oral_preprocessing.Rmd
ps0 <- readRDS("./data/HIV_oral.ps0.gg.rds")
ps0

# Load mapping file
map <- import_qiime_sample_data("./data/mapping_HIV_oral.txt")
ps0 <- merge_phyloseq(ps0, map)

# View sample variables
sample_variables(ps0)

```

Set global ggplot2 theme and options.

```{r global-theme-settings, include=FALSE}
# Set global theming
theme_set(theme_bw(base_size = 10,
                   base_family = "Helvetica"))

```

Factor reordering and renaming

```{r factor-adjustments}
# Factor re-ordering, relabelling, etc.

# Remove the _ from Week 24
sample_data(ps0)$Week
sample_data(ps0)$Week <- factor(sample_data(ps0)$Week, labels = c("Baseline", "Week 24"))
sample_data(ps0)$Week

# Reorder and relabel CD4 category factors
sample_data(ps0)$CD4_category
sample_data(ps0)$CD4_category  <- factor(sample_data(ps0)$CD4_category, levels = c("low", "high"))
sample_data(ps0)$CD4_category

sample_data(ps0)$CD4_category <- factor(sample_data(ps0)$CD4_category, labels = c("CD4 Low", "CD4 High"))
sample_data(ps0)$CD4_category

# Reorder and relabel VL category factors
sample_data(ps0)$VL_category
sample_data(ps0)$VL_category  <- factor(sample_data(ps0)$VL_category, levels = c("low", "high", "Not collected"))
sample_data(ps0)$VL_category

sample_data(ps0)$VL_category <- factor(sample_data(ps0)$VL_category, labels = c("VL Low", "VL High", "NC"))
sample_data(ps0)$VL_category

```

Summary Statistics

```{r data-assessment}
# Create a new data frame of the sorted row sums, a column of sorted values from 1 to the total number of individuals/counts for each RSV and a categorical variable stating these are all RSVs.
readsumsdf = data.frame(nreads = sort(taxa_sums(ps0), TRUE), 
                        sorted = 1:ntaxa(ps0),
                        type = "RSVs")


# Add a column of sample sums (total number of individuals per sample)
readsumsdf = rbind(readsumsdf,
                   data.frame(nreads = sort(sample_sums(ps0), TRUE),
                              sorted = 1:nsamples(ps0),
                              type = "Samples"))

# Make a data frame with a column for the read counts of each sample for histogram production
sample_sum_df <- data.frame(sum = sample_sums(ps0))

# Make plots
# Generates a bar plot with # of reads (y-axis) for each taxa. Sorted from most to least abundant
# Generates a second bar plot with # of reads (y-axis) per sample. Sorted from most to least
p.reads = ggplot(readsumsdf, aes(x = sorted, y = nreads)) +
  geom_bar(stat = "identity") +
  ggtitle("RSV Assessment") +
  scale_y_log10() +
  facet_wrap(~type, scales = "free") +
  ylab("# of Reads")

# Histogram of the number of Samples (y-axis) at various read depths
p.reads.hist <- ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "firebrick3", binwidth = 2500) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  ylab("# of Samples")

# Final plot, side-by-side
grid.arrange(p.reads, p.reads.hist, ncol = 2)

# Basic summary statistics
summary(sample_sums(ps0))

```

Bad sample identification and removal.

```{r sample-removal-identification}
# Format a data table to combine sample summary data with sample variable data
ss <- sample_sums(ps0)
sd <- as.data.frame(sample_data(ps0)) # useful to coerce the phyloseq object into an R data frame
df <- merge(sd, data.frame("RSVs" = ss), by ="row.names") # merge ss with sd by row names. Rename ss to RSVs in the new data frame

# Plot the data by the treatment variable

y = 100 # Set a threshold for the minimum number of acceptable reads. Can start as a guess
x = "Week" # Set the x-axis variable you want to examine
label = "Description" # This is the label you want to overlay on the points that are below threshold y. Should be something sample specific

p.ss.violin <- ggplot(df, aes_string(x, y = "RSVs", fill = x)) + # x is what you assigned it above
  geom_violin() +
  scale_y_log10() +
  geom_hline(yintercept = y, lty = 2) + # Draws a dashed line across the threshold you set above as y
  geom_jitter(alpha = 0.5, width = 0.15) +
  geom_text(data = subset(df, RSVs <= y), aes_string(x, y="RSVs", label=label), size=2) # This labels a subset that fall below threshold variable y and labels them with the label variable

p.ss.violin

# Remove samples with fewer than 100 sequences
ps0 <- prune_samples(sample_sums(ps0)>=100, ps0)
ps0
min(sample_sums(ps0))

```

```{r outlier-sample-evaluation}
# Outlier evaluation
out.bray <- ordinate(ps0, method = "MDS", distance = "jaccard")
p.MDS.outlier <- plot_ordination(ps0, out.bray, color = "Week", axes = c(1,2)) +
  theme_bw() +
  geom_point(size = 2) +
  ggtitle("MDS of Bray Distances \nOutlier Evaluation") +
  stat_ellipse(type = "norm", geom = "polygon", alpha = 1/10, aes(fill = Week))
p.MDS.outlier

```

All samples are realtivaley tighlty clustered. No samples will be removed based on Bray-curtis distances.

```{r prevalence-assessment}
# Begin by removing sequences that were not classified as Bacteria or were classified as either mitochondria or chlorplast

ps0 # Check the number of taxa prior to removal
ps1 <- ps0 %>%
  subset_taxa(
    Kingdom == "Bacteria" &
    Family  != "mitochondria" &
    Class   != "Chloroplast"
  )
ps1 # Confirm that the taxa were removed

# Note the prefix (e.g. k__, f__, c__, etc.) is a GreenGenes convention. If you are using RDP or Silva you can simply remove these prefixes.

# To remove RSVs lacking phyla classification
#ps2 <- subset_taxa(ps2, !is.na(Phylum) & !Phylum %in% c("","uncharacterized"))

# Prevelance estimation
# Calculate feature prevelance across the data set
prevdf <- apply(X = otu_table(ps1),MARGIN = ifelse(taxa_are_rows(ps1), yes = 1, no = 2),FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to prevdf
prevdf <- data.frame(Prevalence = prevdf, TotalAbundance = taxa_sums(ps1), tax_table(ps1))

# Create a table of Phylum, their mean abundances across all samples, and the number of samples they were detected in
plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

#Prevalence plot
prevdf1 <- subset(prevdf, Phylum %in% get_taxa_unique(ps1, "Phylum"))
p.prevdf1 <- ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps1),color=Family)) +
  geom_hline(yintercept = 0.03, alpha = 0.5, linetype = 2) +
  geom_point(size = 3, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) +
  theme(legend.position="none") +
  ggtitle("Phylum Prevelence in All Samples\nColored by Family")
p.prevdf1

```

No reason to filter low-prevelance taxa at this stage.

Data transformations

```{r data-transform}
# Transform to Realative abundances
ps0.ra <- transform_sample_counts(ps0, function(OTU) OTU/sum(OTU))

# Transform to Proportional Abundance
ps0.prop <- transform_sample_counts(ps0, function(x) min(sample_sums(ps0)) * x/sum(x))

# Log transformation moves to a more normal distribution
ps0.log <- transform_sample_counts(ps0, function(x) log(1 + x))

# View how each function altered count data
par(mfrow=c(1,4))
plot(sort(sample_sums(ps0), TRUE), type = "o", main = "Native", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps0.log), TRUE), type = "o", main = "log Transfromed", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps0.ra), TRUE), type = "o", main = "Relative Abundance", ylab = "RSVs", xlab = "Samples")
plot(sort(sample_sums(ps0.prop), TRUE), type = "o", main = "Proportional Abundance", ylab = "RSVs", xlab = "Samples")
par(mfrow=c(1,1))

# Histograms of the non-transformed data vs. the transformed data can address the shift to normality
p.nolog <- qplot(rowSums(otu_table(ps0))) + ggtitle("Raw Counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of Samples")

p.log <- qplot(log10(rowSums(otu_table(ps0)))) +
  ggtitle("log10 transformed counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of Samples")

grid.arrange(p.nolog, p.log, ncol = 2)

```
Subsetting

```{r subsetting}
# Subset. Always subset after taxa/sample filtering
# Baseline
ps0.base <- subset_samples(ps0, Week == "Baseline")
ps0.base
ps0.base <- prune_samples(sample_sums(ps0.base) > 0, ps0.base)
ps0.base
ps0.base.log <- transform_sample_counts(ps0.base, function(x) log(1 + x))

# Week 24
ps0.24 <- subset_samples(ps0, Week == "Week 24")
ps0.24
ps0.24 <- prune_samples(sample_sums(ps0.24) > 0, ps0.24)
ps0.24
ps0.24.log <- transform_sample_counts(ps0.24, function(x) log(1 + x))

# Samples CD4 low at baseline
ps0.CD4.shift <- subset_samples(ps0, CD4_shift_comparison != "exclude")
ps0.CD4.shift
ps0.CD4.shift <- prune_samples(sample_sums(ps0.CD4.shift) > 0, ps0.CD4.shift)
ps0.CD4.shift.log <- transform_sample_counts(ps0.CD4.shift, function(x) log(1 + x))

# Remove Not collected (NC) Viral load samples
ps0
ps0.no_vl_nc <- subset_samples(ps0, VL_category != "NC")
ps0.no_vl_nc

ps0.base
ps0.base.no_vl_nc <- subset_samples(ps0.base, VL_category != "NC")
ps0.base.no_vl_nc

ps0.24
ps0.24.no_vl_nc <- subset_samples(ps0.24, VL_category != "NC")
ps0.24.no_vl_nc

```
##Community composition plotting

```{r community-composition-plots}
# Community composition - Baseline and Week 24 timepoints combined

# Phyla level plots
# Melt to long format (for ggploting) 
# Prune out phyla below % in each sample

ps0_phylum <- ps0 %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt() %>%                                         # Melt to long format
  filter(Abundance > 0.03) %>%                         # Filter out low abundance taxa
  arrange(Phylum)                                      # Sort data frame alphabetically by phylum

# Convert Sample No to a factor because R is weird sometime
ps0_phylum$SampleNo <- as.factor(ps0_phylum$SampleNo)

# Write Phylum table
write.table(ps0_phylum, file = "./results/phylum.txt", sep="\t")

# Plot - Phylum - baseline vs. Week 24
ggplot(ps0_phylum, aes(x = SampleNo, y = Abundance, fill = Phylum)) + 
  theme_bw() +
  geom_bar(stat = "identity") +
  facet_grid(~Week, scales = "free_x") +
  ggtitle("Abundant Phylum (> 3%) in All Samples") +
  ylab("Relative Abundance") +
  theme(axis.text.x = element_blank()) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank()) +
  scale_fill_manual(values = c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e','#e6ab02','#a6761d','#666666'))

```

```{r add-sample-data}
# Diversity
diversity <- global(ps0)
head(diversity)

# Combine diversity into metadata
sample_data(ps0)$Shannon <- diversity$diversities_shannon
sample_data(ps0)$Richness <- diversity$richness_0
colnames(sample_data(ps0))

```

```{r alpha-diversity}
# Make data frame with richness measures
est.rich <- estimate_richness(ps0, measures = c("Observed", "Shannon"))
head(est.rich)

# Make new data table combining richness with sample data
ps0.rich <- cbind(sample_data(ps0), est.rich)
head(ps0.rich)

# log10 transform skewed viral load data
ps0.rich$VL <- log10(ps0.rich$VL)

# CD4 category
p.rich.CD4 <- ggboxplot(ps0.rich, x = "CD4_category", y = "Observed", notch = FALSE, add = "jitter") +
  facet_grid(~Week) +
  stat_compare_means(method = "t.test", label = "p.format")

p.shannon.CD4 <- ggboxplot(ps0.rich, x = "CD4_category", y = "Shannon", notch = FALSE, add = "jitter") +
  facet_grid(~Week) +
  stat_compare_means(method = "t.test", label = "p.format")

grid.arrange(p.rich.CD4, p.shannon.CD4, ncol = 2)

# VL category
# Remove samples with no viral load data
ps0.vl <- subset(ps0.rich, VL_category != "Not collected")

p.rich.VL <- ggboxplot(ps0.vl, x = "VL_category", y = "Observed", notch = FALSE, add = "jitter") +
  facet_grid(~Week) +
  stat_compare_means(method = "t.test", label = "p.format")

p.shannon.VL <- ggboxplot(ps0.vl, x = "VL_category", y = "Shannon", notch = FALSE, add = "jitter") +
  facet_grid(~Week) +
  stat_compare_means(method = "t.test", label = "p.format")

grid.arrange(p.rich.VL, p.shannon.VL, ncol = 2)

grid.arrange(p.rich.CD4, p.shannon.CD4, p.rich.VL, p.shannon.VL, nrow = 2, ncol = 2)

```
Alpha diversity regression

```{r alpha-diversity-regression-CD4}
# The plot_regression command from the microbiome package requires phyloseq objects
ps0.rich.base <- subset(ps0.rich, Week == "Baseline")
ps0.rich.24 <- subset(ps0.rich, Week == "Week 24")

# CD4
p.regr.rich.base.cd4 <- plot_regression(Richness ~ CD4, meta(ps0.rich.base), B=500, show.lm = TRUE, spag = FALSE, shade = FALSE) +
  ggtitle("Richness vs. CD4 (c/ml) Baseline") +
  ylab("Richness") +
  xlab("CD4 (c/ml)") +
  annotate("text", x = 775, y = 325, label = "p = 0.7863") +
  ylim(0,350) +
  xlim(0,900)
cor.test(sample_data(ps0.rich.base)$Richness, sample_data(ps0.rich.base)$CD4)

p.regr.rich.24.cd4 <- plot_regression(Richness ~ CD4, meta(ps0.rich.24), B=500, show.lm = TRUE, spag = FALSE, shade = FALSE) +
  ggtitle("Richness vs. CD4 (c/ml) Week 24") +
  ylab("Richness") +
  xlab("CD4 (c/ml)") +
  annotate("text", x = 775, y = 325, label = "p = 0.0048") +
  ylim(0,350) +
  xlim(0,900)
cor.test(sample_data(ps0.rich.24)$Richness, sample_data(ps0.rich.24)$CD4)

grid.arrange(p.regr.rich.base.cd4, p.regr.rich.24.cd4, ncol = 2)

```

```{r alpha-diversity-regression-VL}
# VL
p.regr.rich.base.vl <- plot_regression(Richness ~ VL, meta(ps0.rich.base), B=500, show.lm = TRUE, spag = FALSE, shade = FALSE) +
  ggtitle("Richness vs. VL (c/ml) Baseline") +
  ylab("Richness") +
  xlab("VL (c/ml)") +
  annotate("text", x = 5.5, y = 325, label = "p = 0.1374") +
  ylim(0,350)
cor.test(sample_data(ps0.rich.base)$Richness, sample_data(ps0.rich.base)$VL)

p.regr.rich.24.vl <- plot_regression(Richness ~ VL, meta(ps0.rich.24), B=500, show.lm = TRUE, spag = FALSE, shade = FALSE) +
  ggtitle("Richness vs. VL (c/ml) Week 24") +
  ylab("Richness") +
  xlab("VL (c/ml)") +
  annotate("text", x = 5.5, y = 325, label = "p = 0.8236") +
  ylim(0,350)
cor.test(sample_data(ps0.rich.24)$Richness, sample_data(ps0.rich.24)$VL)

grid.arrange(p.regr.rich.base.vl, p.regr.rich.24.vl, ncol = 2)

```

```{r ANCOVA-CD4}
# Interaction between age, EE status and alpha diversity
# Or use hotdog to visualize the resutls
ancova(Shannon~CD4*Week, data = ps0.rich)
cor.test(sample_data(ps0.rich.base)$Shannon, sample_data(ps0.rich.base)$CD4)
cor.test(sample_data(ps0.rich.24)$Shannon, sample_data(ps0.rich.24)$CD4)

ancova(Richness~CD4*Week, data = ps0.rich)
cor.test(sample_data(ps0.rich.base)$Richness, sample_data(ps0.rich.base)$CD4)
cor.test(sample_data(ps0.rich.24)$Richness, sample_data(ps0.rich.24)$CD4)

```

```{r ANCOVA-VL}
# Interaction between age, EE status and alpha diversity
# Or use hotdog to visualize the resutls
ancova(Shannon~VL*Week, data = ps0.rich)
cor.test(sample_data(ps0.rich.base)$Shannon, sample_data(ps0.rich.base)$VL)
cor.test(sample_data(ps0.rich.24)$Shannon, sample_data(ps0.rich.24)$VL)

ancova(Richness~VL*Week, data = ps0.rich)
cor.test(sample_data(ps0.rich.base)$Richness, sample_data(ps0.rich.base)$VL)
cor.test(sample_data(ps0.rich.24)$Richness, sample_data(ps0.rich.24)$VL)

```

```{r random-forest}
library(reshape2)
output.forest <- randomForest(CD4_category ~ Shannon + Richness + VL + CD4, na.action = na.omit, data = ps0.rich.base)
print(output.forest)

importance(output.forest, type = 2)
importance.molten <- melt( importance(output.forest, type = 2))

ggplot(importance.molten, aes(x = Var1, y = value)) + geom_bar(stat = "identity")

```
Beta Diversity Analysis

```{r beta-diversity}
# Remove Samples with uncollected (Not Collected) viral loads
sample_data(ps0)$VL_category
ps0.rem_nc <- subset_samples(ps0, VL_category != "Not collected")
sample_data(ps0.rem_nc)$VL_category

ord.ps0.wuni <- ordinate(ps0.rem_nc, method = "PCoA", distance = "wunifrac")

#  Single PCoA lot for presentation / publication
p.ord.CD4 <- plot_ordination(ps0.log.rem_nc, ord.ps0.wuni, color = "Week") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  geom_point(colour = "grey50", size = 1.5) +
  facet_grid(~ CD4_category)

p.ord.VL <- plot_ordination(ps0.log.rem_nc, ord.ps0.wuni, color = "Week") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  geom_point(colour = "grey50", size = 1.5) +
  facet_grid(~ VL_category)

grid.arrange(p.ord.CD4, p.ord.VL, nrow = 2)

 ######### geom_line(aes(group = patient_id), color = "black") +
p.connect <- plot_ordination(ps0.log.rem_nc, ord.ps0.wuni, color = "Week") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  geom_point(colour = "grey50", size = 1.5)
  geom_line(aes(group = patient_id), color = "black")
p.connect

```

#Premanov significance testing

```{r}
# Set a random seed so that exact results can be reproduced
set.seed(1000)

# Function to run adonis test on a physeq object and a variable from metadata 
doadonis <- function(physeq, category) {
  bdist <- phyloseq::distance(physeq, "bray")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col)
  print("Adonis results:")
  print(adonis.bdist)
  
  # Homogeneity of dispersion test
  betatax = betadisper(bdist,col)
  p = permutest(betatax)
  print("Betadisper results:")
  print(p$tab)
}

# Runs Permanov and Homogeneity of dispersion test
# See ?betadisper for more info
doadonis(ps0, "Week")
doadonis(ps0.base, "CD4_category") 
doadonis(ps0.base, "VL_category")

# VL_category at Week 24 has a p < 0.05 (0.049). Test the influence of the Not collected samples on this by removing them and rerunning
doadonis(ps0.24.no_vl_nc, "CD4_category")
doadonis(ps0.24.no_vl_nc, "VL_category")

```

```{r PERMANOVA}
# Permanova test to test significance between groups

set.seed(1000)

# Some recommend to scale to equivalent sample depth so this code is included
# Should test both scaled and unscaled
# Read scaling function
scale_reads <- function(physeq, n) {
  physeq.scale <-
    transform_sample_counts(physeq, function(x) {
      (n * x/sum(x))
    })
  otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}

min.samplesums <- min(sample_sums(ps0)) # Find minimum sample size
ps0_scaled <- scale_reads(ps0, min.samplesums) # Scale to minimum sample size

##### ADONIS ###########

# Function to run adonis test on a physeq object and a variable from metadata 
doadonis <- function(physeq, category) {
  physeq.scale <- scale_reads(physeq, min(sample_sums(physeq)))
  bdist <- phyloseq::distance(physeq.scale, "bray")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col)
  print("Adonis results:")
  print(adonis.bdist)
  
  # Homogeneity of dispersion test
  betatax = betadisper(bdist,col)
  p = permutest(betatax)
  print("Betadisper results:")
  print(p$tab)
}

# Runs Permanov and Homogeneity of dispersion test
# See ?betadisper for more info
doadonis(ps0, "Week")
doadonis(ps0.base.no_vl_nc, "CD4_category") 
doadonis(ps0.base.no_vl_nc, "VL_category")

# VL_category at Week 24 has a p < 0.05 (0.049). Test the influence of the Not collected samples on this by removing them and rerunning
doadonis(ps0.24.no_vl_nc, "CD4_category")
doadonis(ps0.24.no_vl_nc, "VL_category")

```
#Constrained Correspondance Analysis

```{r CCA}
# Initially, all samples were included in the CCA analysis. However, these results clearly show that in the context of CD4 counts and viral load there are two outliers that may skew interpretation. These are the baseline sample from patient 122214 and the week 24 sample from patient 83771. Patient 122214's baseline sample had an extremely low number of CD4 T-cells and a extremely high viral load. Patient 83771's week 24 sample had a very high viral load.
# Due to these factors, these samples were removed from the CCA analysis.

ps0
ps0_cca_out_rm <- ps0 %>%
  subset_samples(
    SampleID != "16750" &
    SampleID != "16767"
  )
ps0_cca_out_rm

# Constrained Ordination - Baseline

# Remove data points with missing metadata
ps0_not_na <- ps0_cca_out_rm%>%
  subset_samples(
    !is.na(CD4) &
    !is.na(VL)
      )

# Subset and log transforme not_na samples
ps0_not_na.log <- transform_sample_counts(ps0_not_na, function(x) log(1 + x))
ps0.not_na.base <- subset_samples(ps0_not_na.log, Week == "Baseline")
ps0.not_na.base
ps0.not_na.24 <- subset_samples(ps0_not_na.log, Week == "Week 24")
ps0.not_na.24

# Calculate distance
bray_not_na.base <- phyloseq::distance(physeq = ps0.not_na.base, method = "unifrac")
bray_not_na.24 <- phyloseq::distance(physeq = ps0.not_na.24, method = "unifrac")

# CAP Ordinate
cap_ord.base <- ordinate(ps0.not_na.base, method = "CAP", distance = bray_not_na.base, formula = ~VL + CD4)
cap_ord.24 <- ordinate(ps0.not_na.24, method = "CAP", distance = bray_not_na.24, formula = ~VL + CD4)

# CAP plot - Baseline
cap_plot.base <- plot_ordination(ps0.not_na.base, ordination = cap_ord.base) +
  theme_bw() +
  geom_point(aes(colour = CD4_category), alpha = 0.7, size = 4) +
  geom_point(colour = "grey50", size = 1.5) +
  scale_color_brewer(palette = "Dark2") +
  ggtitle("CCA ~ CD4 T Cells + Viral Load \n Baseline") +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(colour = "CD4 Category")
  #geom_text(aes(label = patient_id), size = 3, check_overlap = TRUE, vjust = 2)

# CAP plot - Week 24
cap_plot.24 <- plot_ordination(ps0.not_na.24, ordination = cap_ord.24) +
  theme_bw() +
  geom_point(aes(colour = CD4_category), alpha = 0.7, size = 4) +
  geom_point(colour = "grey50", size = 1.5) +
  scale_color_brewer(palette = "Dark2") +
  ggtitle("CCA ~ CD4 T Cells + Viral Load \n Week 24") +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(colour = "CD4 Category")
  # geom_text(aes(label = patient_id), size = 3, check_overlap = TRUE, vjust = 2)

# Now add the environmental variables as arrows
arrowmat.base <- vegan::scores(cap_ord.base, display = "bp")
arrowmat.24 <- vegan::scores(cap_ord.24, display = "bp")

# Add labels, make a data.frame
arrowdf.base <- data.frame(labels = rownames(arrowmat.base), arrowmat.base)
arrowdf.24 <- data.frame(labels = rownames(arrowmat.24), arrowmat.24)

# Define the arrow aesthetic mapping
arrow_map <- aes(xend = CAP1, 
    yend = CAP2, 
    x = 0, 
    y = 0, 
    shape = NULL, 
    color = NULL, 
    label = labels)

label_map <- aes(x = 1.3 * CAP1, 
    y = 1.3 * CAP2, 
    shape = NULL, 
    color = NULL, 
    label = labels)

arrowhead = arrow(length = unit(0.02, "npc"))

# Make a new graphic
# CAP plot with arrows - baseline
cap_plot.base_arrows <- cap_plot.base + 
  geom_segment(
    mapping = arrow_map, 
    size = .5, 
    data = arrowdf.base, 
    color = "gray", 
    arrow = arrowhead
  ) + 
  geom_text(
    mapping = label_map, 
    size = 4,  
    data = arrowdf.base, 
    show.legend = FALSE
  )

# CAP plot with arrows - Week 24
cap_plot.24_arrows <- cap_plot.24 + 
  geom_segment(
    mapping = arrow_map, 
    size = .5, 
    data = arrowdf.24, 
    color = "gray", 
    arrow = arrowhead
  ) + 
  geom_text(
    mapping = label_map, 
    size = 4,  
    data = arrowdf.24, 
    show.legend = FALSE
  )

grid.arrange(cap_plot.base_arrows, cap_plot.24_arrows, ncol = 2)

anova(cap_ord.base, permutations = 9999)
anova(cap_ord.24, permutations = 9999) 

```
#Differential Abundance Testing - Week

```{r diffab-baseline-vs-week-24}
# Differential Abundance - Week
ds.week <- phyloseq_to_deseq2(ps0, ~Week)
dds.week <- DESeq(ds.week, test="Wald", fitType="local", betaPrior = FALSE) # Worth trying parametric and local fitTypes
res.dds.week = results(dds.week, cooksCutoff = FALSE)
alpha = 0.1 # Note: this is higher to try and allow even weak effect size taxa through
sigtab_dds.week = res.dds.week[which(res.dds.week$padj < alpha), ]
head(sigtab_dds.week) # Empty table

# There are no taxa differentially abundant at p = 0.1 between Week 0 and Week 24

```
#Differential Abundance Testing - Baseline Samples

```{r diffab-baseline-samples}
# Differential Abundance - Baseline ~ CD4 and VL Category
ds.base <- phyloseq_to_deseq2(ps0.base.no_vl_nc, ~CD4_category) # Change formula to VL_category and rerun to test VL_category
dds.base <- DESeq(ds.base, test="Wald", fitType="local", betaPrior = FALSE) # Worth trying parametric and local fitTypes

# Tabulate and write results
alpha = 0.1
res.dds.base = results(dds.base, cooksCutoff = FALSE)
sigtab_dds.base = res.dds.base[which(res.dds.base$padj < alpha), ]
sigtab_dds.base = cbind(as(sigtab_dds.base, "data.frame"), as(tax_table(ps0.base.no_vl_nc)[rownames(sigtab_dds.base), ], "matrix"))
head(sigtab_dds.base)
write.table(sigtab_dds.base, file="./results/deseq_results_base_CD4.txt", sep = "\t") # Change filename to save results to appropriate file

x.ds.base = tapply(sigtab_dds.base$log2FoldChange, sigtab_dds.base$Family, function(x) max(x))
x.ds.base = sort(x.ds.base, TRUE)
sigtab_dds.base$Species = factor(as.character(sigtab_dds.base$Family), levels=names(x.ds.base))
p.deseq.base <- ggplot(sigtab_dds.base, aes(x=Family, y=log2FoldChange, color=Genus)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_point(size=5, alpha = 0.7) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  ggtitle("Taxa Differentially Associated by XXXX Category \n Baseline") + # Change title if variable changed
  geom_hline(yintercept = 0, lwd=1.5)
p.deseq.base

# Quick check of factor levels
mcols(res.dds.base, use.names = TRUE)

# Create volcano plots of signficant results
# Prepare to join results data.table and taxonomy table
# Baseline
resdt.base = data.table(as(results(dds.base, cooksCutoff = FALSE), "data.frame"),
                   keep.rownames = TRUE)
setnames(resdt.base, "rn", "OTU")
taxdt.base = data.table(data.frame(as(tax_table(ps0.base), "matrix")), keep.rownames = TRUE)
setnames(taxdt.base, "rn", "OTU")

# Join results data.table and taxonomy table
setkeyv(taxdt.base, "OTU")
setkeyv(resdt.base, "OTU")
resdt.base <- taxdt.base[resdt.base]
resdt.base
resdt.base[, Significant := padj < alpha]
resdt.base[!is.na(Significant)]
resdt.base

volcano.base = ggplot(
  data = resdt.base[!is.na(Significant)][(pvalue < 1)],
  mapping = aes(x = log2FoldChange,
                y = -log10(pvalue),
                color = Phylum,
                label = OTU, label1 = Genus)) +
  theme_bw() +
  geom_point() + 
  geom_point(data = resdt.base[(Significant)], size = 7, alpha = 0.7) + 
  # geom_text(data = resdt.base[(Significant)], mapping = aes(label = paste("Genus:", Genus)), color = "black", size = 3) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  geom_hline(yintercept = -log10(alpha)) +
  ggtitle("DESeq2 Negative Binomial Test Volcano Plot\nBaseline Samples") +
  theme(axis.title = element_text(size=12)) +
  theme(axis.text = element_text(size=12)) +
  theme(legend.text = element_text(size=12)) +
  geom_vline(xintercept = 0, lty = 2)
volcano.base

ggplotly(volcano.base)

```
#Differential Abundance Testing - Week 24 Samples

```{r diffab-week-24-samples}
# Differential Abundance - 24line ~ CD4 and VL Category
ds.24 <- phyloseq_to_deseq2(ps0.24.no_vl_nc, ~CD4_category) # Change formula to VL_category and rerun to test VL_category
dds.24 <- DESeq(ds.24, test="Wald", fitType="local", betaPrior = FALSE) # Worth trying parametric and local fitTypes

res.dds.24 = results(dds.24, cooksCutoff = FALSE)
alpha = 0.1
sigtab_dds.24 = res.dds.24[which(res.dds.24$padj < alpha), ]
head(sigtab_dds.24)

# There are no taxa differentially abundant at p = 0.1 between CD4 or VL categories at Week 24

```

```{r diffab-shift}
# Subset samples that were CD4 low at baseline and shifted to high at week 24
# All of these samples had VL taken at baseline and Week 24, so no concern for samples that are missing VL data
ps0
ps0.shift <- subset_samples(ps0, low_CD4_baseline_24wk_sample != "No")
ps0.shift
sample_data(ps0.shift)$low_CD4_baseline_24wk_sample

# Differential Abundance - shiftline ~ CD4 and VL Category
ds.shift <- phyloseq_to_deseq2(ps0.shift, ~low_CD4_baseline_24wk_sample) # Change formula to VL_category and rerun to test VL_category
dds.shift <- DESeq(ds.shift, test="Wald", fitType="local", betaPrior = FALSE) # Worth trying parametric and local fitTypes

res.dds.shift = results(dds.shift, cooksCutoff = FALSE)
alpha = 0.1
sigtab_dds.shift = res.dds.shift[which(res.dds.shift$padj < alpha), ]
sigtab_dds.shift = cbind(as(sigtab_dds.shift, "data.frame"), as(tax_table(ps0.shift)[rownames(sigtab_dds.shift), ], "matrix"))
head(sigtab_dds.shift)
write.table(sigtab_dds.shift, file="./results/deseq_results_shift.txt", sep = "\t") # Change filename to save results to appropriate file

x.ds.shift = tapply(sigtab_dds.shift$log2FoldChange, sigtab_dds.shift$Family, function(x) max(x))
x.ds.shift = sort(x.ds.shift, TRUE)
sigtab_dds.shift$Species = factor(as.character(sigtab_dds.shift$Family), levels=names(x.ds.shift))
p.deseq.shift <- ggplot(sigtab_dds.shift, aes(x=Family, y=log2FoldChange, color=Genus)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_point(size=5, alpha = 0.7) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  ggtitle("Taxa Differentially Associated by XXXX Category \n shiftline") + # Change title if variable changed
  geom_hline(yintercept = 0, lwd=1.5)
p.deseq.shift

# Quick check of factor levels
mcols(res.dds.shift, use.names = TRUE)

# Create volcano plots of signficant results
# Prepare to join results data.table and taxonomy table
# shiftline
resdt.shift = data.table(as(results(dds.shift, cooksCutoff = FALSE), "data.frame"),
                   keep.rownames = TRUE)
setnames(resdt.shift, "rn", "OTU")
taxdt.shift = data.table(data.frame(as(tax_table(ps0.shift), "matrix")), keep.rownames = TRUE)
setnames(taxdt.shift, "rn", "OTU")

# Join results data.table and taxonomy table
setkeyv(taxdt.shift, "OTU")
setkeyv(resdt.shift, "OTU")
resdt.shift <- taxdt.shift[resdt.shift]
resdt.shift
resdt.shift[, Significant := padj < alpha]
resdt.shift[!is.na(Significant)]
resdt.shift

volcano.shift = ggplot(
  data = resdt.shift[!is.na(Significant)][(pvalue < 0.9)],
  mapping = aes(x = log2FoldChange,
                y = -log10(pvalue),
                color = Phylum,
                label = OTU, label1 = Genus)) +
  theme_bw() +
  geom_point() + 
  geom_point(data = resdt.shift[(Significant)], size = 7, alpha = 0.7) + 
  # geom_text(data = resdt.shift[(Significant)], mapping = aes(label = paste("Genus:", Genus)), color = "black", size = 3) +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  geom_hline(yintercept = -log10(alpha)) +
  ggtitle("DESeq2 Negative Binomial Test Volcano Plot\nshiftline Samples") +
  theme(axis.title = element_text(size=12)) +
  theme(axis.text = element_text(size=12)) +
  theme(legend.text = element_text(size=12))
volcano.shift

ggplotly(volcano.shift)

```
#Random Forest

```{r random-forest}
# Random Forest
# Make a dataframe of training data with OTUs as column and samples as rows
predictors.base <- otu_table(ps0.base)
dim(predictors.base)

# Make one column for our outcome/response variable 
response.base <- as.factor(sample_data(ps0.base)$CD4_category)

# Combine them into 1 data frame
rf.data <- data.frame(response.base, predictors.base)

# Results
set.seed(2)
classify <- randomForest(response.base~., data = rf.data, ntree = 100)
print(classify)

# Make a data frame with predictor names and their importance
imp <- importance(classify)
imp <- data.frame(predictors = rownames(imp), imp)

# Order the predictor levels by importance
imp.sort <- arrange(imp, desc(MeanDecreaseGini))
imp.sort$predictors <- factor(imp.sort$predictors, levels = imp.sort$predictors)

# Select the top 10 predictors
imp.20 <- imp.sort[1:20, ]

# ggplot
ggplot(imp.20, aes(x = predictors, y = MeanDecreaseGini)) +
  geom_bar(stat = "identity", fill = "indianred") +
  theme(axis.text.x = element_blank())
  coord_flip()

# What are those OTUs?
otunames <- imp.20$predictors
r <- rownames(tax_table(ps0.base)) %in% otunames
kable(tax_table(ps0.base)[r, ])

```

Taxon specific barplots from DA taxa identified with DeSeq2.

```{r ground-truth}
# Plots of DeSeq2 identified taxa

# Subset DeSeq2 identified taxa
# Can choose between 2 transformations from above
# Typically go with relative abundances as these are more familiar to readers

# Genus = Capnocytophaga
ps0_Capnocytophaga <- ps0.no_vl_nc.ra %>%
    subset_taxa(Genus == "Capnocytophaga") %>%
    psmelt() 

# Genus = Aggregatibacter
ps0_Aggregatibacter <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Aggregatibacter") %>%
  psmelt()

# Family = [Paraprevotellaceae]
ps0_Paraprevotellaceae <- ps0.no_vl_nc.ra %>%
  subset_taxa(Family == "[Paraprevotellaceae]") %>%
  psmelt()

# Genus = Dialister
ps0_Dialister <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Dialister") %>%
  psmelt()

# Genus = Fusobacterium
ps0_Fusobacterium <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Fusobacterium") %>%
  psmelt()

# Genus = Veillonella
ps0_Veillonella <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Veillonella") %>%
  psmelt()

# Genus = Streptococcus
ps0_Streptococcus <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Streptococcus") %>%
  psmelt()

# Genus = Neisseria
ps0_Neisseria <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Neisseria") %>%
  psmelt()

# Genus = Prevotella
ps0_Prevotella <- ps0.no_vl_nc.ra %>%
  subset_taxa(Genus == "Prevotella") %>%
  psmelt()
  
# Capnocytophaga violin-point plots
p.Capnocytophaga.VL <- ggplot(ps0_Capnocytophaga, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Capnocytophaga.VL

# Aggregatibacter violin-point plots
p.Aggregatibacter.VL <- ggplot(ps0_Aggregatibacter, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Aggregatibacter.VL

# [Paraprevotellaceae] violin-point plots
p.Paraprevotellaceae.VL <- ggplot(ps0_Paraprevotellaceae, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Paraprevotellaceae.VL

# Dialister violin-point plots
p.Dialister.VL <- ggplot(ps0_Dialister, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Dialister.VL

# Fusobacterium violin-point plots
p.Fusobacterium.VL <- ggplot(ps0_Fusobacterium, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Fusobacterium.VL

# Veillonella violin-point plots
p.Veillonella.VL <- ggplot(ps0_Veillonella, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey") +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Veillonella.VL

# Streptococcus violin-point plots
p.Streptococcus.VL <- ggplot(ps0_Streptococcus, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = FALSE, trim = FALSE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Streptococcus.VL

# Neisseria violin-point plots
p.Neisseria.VL <- ggplot(ps0_Neisseria, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = FALSE, trim = FALSE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Neisseria.VL

# Prevotella violin-point plots
p.Prevotella.VL <- ggplot(ps0_Prevotella, aes(x = VL_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = FALSE, trim = FALSE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Prevotella.VL

```

```{r DeSeq2 Identified Taxa by CD4 Category, fig.width=9, fig.height=6}
# Same plots as above, but for CD4_category
# Capnocytophaga violin-point plots
p.Capnocytophaga.CD4 <- ggplot(ps0_Capnocytophaga, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Capnocytophaga.CD4

# Aggregatibacter violin-point plots
p.Aggregatibacter.CD4 <- ggplot(ps0_Aggregatibacter, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Aggregatibacter.CD4

# [Paraprevotellaceae] violin-point plots
p.Paraprevotellaceae.CD4 <- ggplot(ps0_Paraprevotellaceae, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Paraprevotellaceae.CD4

# Dialister violin-point plots
p.Dialister.CD4 <- ggplot(ps0_Dialister, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Dialister.CD4

# Fusobacterium violin-point plots
p.Fusobacterium.CD4 <- ggplot(ps0_Fusobacterium, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = TRUE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Fusobacterium.CD4

# Veillonella violin-point plots
p.Veillonella.CD4 <- ggplot(ps0_Veillonella, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey") +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Veillonella.CD4

# Streptococcus violin-point plots
p.Streptococcus.CD4 <- ggplot(ps0_Streptococcus, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = FALSE, trim = FALSE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Streptococcus.CD4

# Neisseria violin-point plots
p.Neisseria.CD4 <- ggplot(ps0_Neisseria, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = FALSE, trim = FALSE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Neisseria.CD4

# Prevotella violin-point plots
p.Prevotella.CD4 <- ggplot(ps0_Prevotella, aes(x = CD4_category, y = Abundance, fill = Genus)) + 
  theme_bw() +
  scale_size_continuous(range = c(2, 9)) +
  geom_violin(na.rm = TRUE, alpha = 0.5, fill = "grey", draw_quantiles = FALSE, trim = FALSE) +
  facet_grid(Week~Species, scales = "free_x", space = "free_x") +
  scale_y_log10() +
  geom_point(mapping = aes(size = sqrt(Abundance), color = Species), alpha = 0.4, na.rm = TRUE) +
  theme(strip.text = element_text(size = 12)) +
  theme(axis.text = element_text(size = 12)) +
  theme(axis.title = element_text(size = 12)) +
  theme(legend.text = element_text(size = 12)) +
  theme(axis.title.x = element_blank())
p.Prevotella.CD4

```

