---
title: "HIV - Oral - Preprocessing"
author: "Scott A. Handley"
date: "1/3/2017"
output: html_document
---

dada2 analysis of WNV infected Vancomycin treated mice.

References:
  http://f1000research.com/articles/5-1492/v1
  http://benjjneb.github.io/dada2/tutorial.html

```{r}
# Load libraries
library("ShortRead")
packageVersion("ShortRead")
library(dada2)
packageVersion("dada2")
library(ggplot2)
packageVersion("ggplot2")
library("msa")
packageVersion("msa")
library("phangorn")
packageVersion("phangorn")
library("phyloseq")
packageVersion("phyloseq")

```

```{r}
# Set file paths
path <- "~/Desktop/HIV_oral_dada2/"

# Create ordered list of files
fns <- sort(list.files(path))
fns

```

```{r}
# View sequence quality

# Isolate the fastq files just in case you have other file types in the folder
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs) # Sort ensures forward/reverse reads are in same order. May or may not be necessary
fnFs <- fastqs[grepl("_R1", fastqs)] # Just the forward read files. Note: Must have _R1 in file name.
fnFs
fnRs <- fastqs[grepl("_R2", fastqs)] # Just the reverse read files. Note: Must have _R2 in file name.
fnRs

# Get sample names from the first part of the forward read filenames
sample.names <- sapply(strsplit(fnFs, "_"), `[`, 1) # Note. This just pulls everything before the _.
sample.names

# Fully specify the path for the fnFs and fnRs
fnFs <- file.path(path, fnFs)
fnFs
fnRs <- file.path(path, fnRs)
fnRs

# Plot forward and reverse read quality
plotQualityProfile(fnFs[1]) + ggtitle("Fwd")
plotQualityProfile(fnRs[1]) + ggtitle("Rev")

# Read 2 is very low-quality

```

```{r}
# Make directory and filenames for the filtered fastqs
filt_path <- file.path(path, "filtered")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))
# Filter
for(i in seq_along(fnFs)) {
  fastqPairedFilter(c(fnFs[i], fnRs[i]), c(filtFs[i], filtRs[i]),
                    trimLeft=c(10, 10), truncLen=c(200,200), 
                    maxN=0, maxEE=2, truncQ=2, 
                    compress=TRUE, verbose=TRUE, rm.phix =c(TRUE, TRUE))
}

# Probably just going to analyze R1 due to lower quality and questionable value of R2

```

```{r}
# Dereplicate sequences
derepFs <- derepFastq(filtFs, verbose=TRUE)

# Name the derep-class objects by the sample names
names(derepFs) <- sample.names

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```

```{r}
# Sequence inference
dadaFs <- dada(derepFs, err=NULL, selfConsist = TRUE, pool = TRUE, multithread = TRUE)

plotErrors(dadaFs, nominalQ = TRUE) + ggtitle("Forward")

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```

```{r}
#mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
#head(mergers[[1]])

```

```{r}
# Construct sequence table
seqtab <- makeSequenceTable(dadaFs[names(dadaFs) != "Mock"])
dim(seqtab)
write.table(seqtab, file = "~/Desktop/HIV_oral_dada2/seqtab_full_HIV_oral.txt", sep = "\t")

# Inspect distribution of sequence lengths
table(nchar(colnames(seqtab)))

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```

```{r}
# Remove chimeras
seqtab.nochim <- removeBimeraDenovo(seqtab, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```

```{r}
## Assign Taxonomy
# GreenGenes
taxa.gg <- assignTaxonomy(seqtab.nochim, "~/Desktop/FGT_dada2/gg_13_8_train_set_97.fa.gz")
unname(head(taxa.gg))
colnames(taxa.gg) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

# RDP
taxa.rdp <- assignTaxonomy(seqtab.nochim, "~/Desktop/FGT_dada2/rdp_train_set_14.fa.gz")
unname(head(taxa.rdp))
colnames(taxa.rdp) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")

# Silva
taxa.silva <- assignTaxonomy(seqtab.nochim, "~/Desktop/FGT_dada2/silva_nr_v123_train_set.fa.gz")
unname(head(taxa.silva))
colnames(taxa.silva) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")

# Add species to RDP and Silva
taxa.rdp.plus <- addSpecies(taxa.rdp, "~/Desktop/dada2_species_databases/assignTaxonomy/rdp_species_assignment_14.fa.gz", verbose = TRUE)
unname(head(taxa.rdp.plus))
colnames(taxa.rdp.plus) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

# 218 out of 1947 were assigned to the species level.
# Of which 212 had genera consistent with the input table.

# SILVA
taxa.silva.plus <- addSpecies(taxa.silva, "~/Desktop/dada2_species_databases/assignTaxonomy/silva_species_assignment_v123.fa.gz", verbose = TRUE)
unname(head(taxa.silva.plus))
colnames(taxa.silva.plus) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

# 146 out of 601 were assigned to the species level.
# Of which 142 had genera consistent with the input table.

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```

```{r}
# Construct the phylogenetic tree
seqs <- getSequences(seqtab.nochim)

names(seqs) <- seqs # This propagates to the tip labels of the tree
mult <- msa(seqs, method="ClustalW", type="dna", order="input")

phang.align <- as.phyDat(mult, type="DNA", names=getSequence(seqtab.nochim))

dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phang.align)

fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                       rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```

```{r}
# Create PhyloSeq objects
# Greengenes
# Strip the r__ prefixes from taxon labels. This is only needed for the GreenGenes taxon annotations
taxa.gg.fixed <- gsub("k__", "", taxa.gg)
taxa.gg.fixed <- gsub("p__", "", taxa.gg.fixed)
taxa.gg.fixed <- gsub("c__", "", taxa.gg.fixed)
taxa.gg.fixed <- gsub("o__", "", taxa.gg.fixed)
taxa.gg.fixed <- gsub("f__", "", taxa.gg.fixed)
taxa.gg.fixed <- gsub("g__", "", taxa.gg.fixed)
taxa.gg.fixed <- gsub("s__", "", taxa.gg.fixed)

ps0.gg <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), tax_table(taxa.gg.fixed), phy_tree(fitGTR$tree))
ps0.gg

# RDP
ps0.rdp <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), tax_table(taxa.rdp.plus), phy_tree(fitGTR$tree))
ps0.rdp

# Silva
ps0.silva <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), tax_table(taxa.silva.plus), phy_tree(fitGTR$tree))
ps0.silva

# Import mapping data
# Mapping fileikely has updates since OTU picking
# This enables import of updated mapping data
map.dada <- import_qiime_sample_data("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/data/mapping_HIV_oral.txt")
dim(map.dada)
head(map.dada)

# Sanity checks
dim(map.dada)

# Merge OTU table with mapping data
ps0.gg <- merge_phyloseq(ps0.gg, map.dada)
ps0.gg

ps0.rdp <- merge_phyloseq(ps0.rdp, map.dada)
ps0.rdp

ps0.silva <- merge_phyloseq(ps0.silva, map.dada)
ps0.silva

# Sanity checks
rank_names(ps0.gg)
nsamples(ps0.gg)
sample_variables(ps0.gg)
ntaxa(ps0.gg)
taxa_names(ps0.gg)[1:10]
get_taxa_unique(ps0.gg, "Phylum")

# Basic read depth statistics
min(sample_sums(ps0.gg))
max(sample_sums(ps0.gg))
mean(sample_sums(ps0.gg))
median(sample_sums(ps0.gg))

# Save Phyloseq Object as RDS
saveRDS(ps0.gg, "~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/data/HIV_oral.ps0.gg.rds")
saveRDS(ps0.rdp, "~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/data/HIV_oral.ps0.rdp.rds")
saveRDS(ps0.silva, "~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/data/HIV_oral.ps0.silva.rds")

save.image("~/Desktop/HIV_oral_dada2/HIV_oral_dada2.RData")
save.image("~/Dropbox/Research/AIDS/HIV/presti/HIV_oral/dada2/HIV_oral_dada2.RData")

```
